pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/array/ListAggFunction.java

```diff
@@ -0,0 +1,214 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.query.aggregation.function.array;
+
+import it.unimi.dsi.fastutil.objects.ObjectArrayList;
+import java.util.Map;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.pinot.common.request.context.ExpressionContext;
+import org.apache.pinot.common.utils.DataSchema;
+import org.apache.pinot.core.common.BlockValSet;
+import org.apache.pinot.core.query.aggregation.AggregationResultHolder;
+import org.apache.pinot.core.query.aggregation.ObjectAggregationResultHolder;
+import org.apache.pinot.core.query.aggregation.function.BaseSingleInputAggregationFunction;
+import org.apache.pinot.core.query.aggregation.groupby.GroupByResultHolder;
+import org.apache.pinot.core.query.aggregation.groupby.ObjectGroupByResultHolder;
+import org.apache.pinot.segment.spi.AggregationFunctionType;
+import org.roaringbitmap.RoaringBitmap;
+
+
+public class ListAggFunction
+    extends BaseSingleInputAggregationFunction<ObjectArrayList<String>, String> {
+
+  private final String _separator;
+  private final boolean _nullHandlingEnabled;
+  private final boolean _isDistinct;
+
+  public ListAggFunction(ExpressionContext expression, String separator, boolean isDistinct,
+      boolean nullHandlingEnabled) {
+    super(expression);
+    _separator = separator;
+    _isDistinct = isDistinct;
+    _nullHandlingEnabled = nullHandlingEnabled;
+  }
+
+  @Override
+  public AggregationFunctionType getType() {
+    return AggregationFunctionType.LISTAGG;
+  }
+
+  @Override
+  public AggregationResultHolder createAggregationResultHolder() {
+    return new ObjectAggregationResultHolder();
+  }
+
+  @Override
+  public GroupByResultHolder createGroupByResultHolder(int initialCapacity, int maxCapacity) {
+    return new ObjectGroupByResultHolder(initialCapacity, maxCapacity);
+  }
+
+  @Override
+  public void aggregate(int length, AggregationResultHolder aggregationResultHolder,
+      Map<ExpressionContext, BlockValSet> blockValSetMap) {
+    ObjectArrayList<String> valueSet = aggregationResultHolder.getResult();
+    if (valueSet == null) {
+      valueSet = new ObjectArrayList<>();
+      aggregationResultHolder.setValue(valueSet);
+    }
+    BlockValSet blockValSet = blockValSetMap.get(_expression);
+    String[] values = blockValSet.getStringValuesSV();
+    if (_nullHandlingEnabled) {
+      RoaringBitmap nullBitmap = blockValSet.getNullBitmap();
+      if (nullBitmap != null && !nullBitmap.isEmpty()) {
+        aggregateArrayWithNull(length, valueSet, values, nullBitmap);
+        return;
+      }
+    }
+    aggregateArray(length, valueSet, values);
+  }
+
+  private void aggregateArray(int length, ObjectArrayList<String> valueSet, String[] values) {
+    for (int i = 0; i < length; i++) {
+      String value = values[i];
+      if (_isDistinct && valueSet.contains(value)) {
```

Comment:
This is `O(n)` contains, which will be extremely slow

Reply:
there is no good way for this, unless we keep two copies, one is the array list for ordering, the other is a hash map for collision checks